### PR TITLE
Making the process implementation simpler

### PR DESCRIPTION
**Pull request by roynalnaruto**

### CODE REVIEW

File: replica/replica.go [hunk 1]
```diff
@@ -60,81 +73,153 @@ func New(opts Options, whoami id.Signatory, signatories []id.Signatory, propose
...
+
+			select {
+			case <-ctx.Done():
+				isRunning = false
+				return
+
+			case timeout := <-replica.onProposeTimeout:
+				if !replica.filterHeight(timeout.Height) {
+					return
+				}
```
Comment (loongy) [inline replica/replica.go:110]: These are self-generated messages, so we should not filter by height. The `Process` logic already takes care of timeouts that are from the "wrong" height. 
```suggestion
```

File: replica/replica.go [hunk 2]
```diff
@@ -60,81 +73,153 @@ func New(opts Options, whoami id.Signatory, signatories []id.Signatory, propose
...
+				replica.proc.OnTimeoutPropose(timeout.Height, timeout.Round)
+			case timeout := <-replica.onPrevoteTimeout:
+				if !replica.filterHeight(timeout.Height) {
+					return
+				}
```
Comment (loongy) [inline replica/replica.go:115]:
```suggestion
```

File: replica/replica.go [hunk 3]
```diff
@@ -60,81 +73,153 @@ func New(opts Options, whoami id.Signatory, signatories []id.Signatory, propose
...
+				replica.proc.OnTimeoutPrevote(timeout.Height, timeout.Round)
+			case timeout := <-replica.onPrecommitTimeout:
+				if !replica.filterHeight(timeout.Height) {
+					return
+				}
```
Comment (loongy) [inline replica/replica.go:120]:
```suggestion
```

File: timer/timer.go [hunk 4]
```diff
@@ -1,50 +1,100 @@
...
 	}
 }
 
+// TimeoutPropose schedules a propose timeout with a timeout period appropriately
+// calculated for the consensus height and round
 func (t *LinearTimer) TimeoutPropose(height process.Height, round process.Round) {
 	go func() {
 		time.Sleep(t.timeoutDuration(height, round))
-		t.onTimeoutPropose <- Timeout{Height: height, Round: round}
+		t.handleTimeoutPropose(Timeout{Height: height, Round: round})
```
Comment (loongy) [inline timer/timer.go:79]: Allow the handler to be nil.
```suggestion
		if t.handleTimeoutPropose != nil {
			t.handleTimeoutPropose(Timeout{Height: height, Round: round})
		}
```

File: timer/timer.go [hunk 5]
```diff
@@ -1,50 +1,100 @@
...
 	}()
 }
 
+// TimeoutPrevote schedules a prevote timeout with a timeout period appropriately
+// calculated for the consensus height and round
 func (t *LinearTimer) TimeoutPrevote(height process.Height, round process.Round) {
 	go func() {
 		time.Sleep(t.timeoutDuration(height, round))
-		t.onTimeoutPrevote <- Timeout{Height: height, Round: round}
+		t.handleTimeoutPrevote(Timeout{Height: height, Round: round})
```
Comment (loongy) [inline timer/timer.go:88]:
```suggestion
		if t.handleTimeoutPrevote != nil {
			t.handleTimeoutPrevote(Timeout{Height: height, Round: round})
		}
```

File: timer/timer.go [hunk 6]
```diff
@@ -1,50 +1,100 @@
...
 	}()
 }
 
+// TimeoutPrecommit schedules a precommit timeout with a timeout period appropriately
+// calculated for the consensus height and round
 func (t *LinearTimer) TimeoutPrecommit(height process.Height, round process.Round) {
 	go func() {
 		time.Sleep(t.timeoutDuration(height, round))
-		t.onTimeoutPrecommit <- Timeout{Height: height, Round: round}
+		t.handleTimeoutPrecommit(Timeout{Height: height, Round: round})
```
Comment (loongy) [inline timer/timer.go:97]:
```suggestion
		if t.handleTimeoutPrecommit != nil {
			t.handleTimeoutPrecommit(Timeout{Height: height, Round: round})
		}
```

File: timer/timer.go (hunk 4 above)
Comment (roynalnaruto) [reply timer/timer.go:79]: In that case I will simply wrap the entire `go func() { ... }()` by the `nil` check? If the handler is `nil`, then we can probably avoid the `time.Sleep` as well?

**Pull request merged**